xds/src/main/java/io/grpc/xds/XdsClientWrapperForServerSds.java

```diff
@@ -0,0 +1,281 @@
+        }
+      }
+      if (bestMatch != null) {
+        return bestMatch.getDownstreamTlsContext();
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Computes a score for a match of filterChainMatch with channel.
+   *
+   * <p>-1 => mismatch (of port or IP address etc)
+   *
+   * <p>0 => unimplemented (types or logic etc)
+   *
+   * <p>1 => filterChainMatch is null, so nothing to match.
+   *
+   * <p>value > 2 indicates some kind of IP address match.
+   */
+  private static int getFilterChainMatchScore(
+      FilterChainMatch filterChainMatch, InetSocketAddress localInetAddr) {
+    if (filterChainMatch == null) {
+      return 1;
+    }
+    if (filterChainMatch.getDestinationPort() != localInetAddr.getPort()) {
+      return -1;
+    }
+    return getIpAddressAndRangesMatchScore(
+        localInetAddr.getAddress(), filterChainMatch.getPrefixRanges());
+  }
+
+  /**
+   * Computes a score for IP address match against a list of CidrRange called only after ports have
+   * matched, so a minimum score of 2 is returned to indicate a match or else -1 for a mismatch.
+   *
+   * <p>-1 => mismatch.
+   *
+   * <p>2 => prefixRanges is null/empty, so no IP address to match.
+   *
+   * <p>value > 2 indicates some kind of IP address match.
+   */
+  private static int getIpAddressAndRangesMatchScore(
+      InetAddress localAddress, List<CidrRange> prefixRanges) {
+    if (prefixRanges == null || prefixRanges.isEmpty()) {
+      return 2;
+    }
+    int highestScore = -1;
+    for (CidrRange cidrRange : prefixRanges) {
+      int curScore = getIpAddressMatchScore(localAddress, cidrRange);
+      if (curScore > highestScore) {
+        highestScore = curScore;
+      }
+    }
+    return highestScore;
+  }
+
+  /**
+   * Computes a score for IP address to CidrRange match.
+   *
+   * <p>-1 => mismatch
+   *
+   * <p>0 => unimplemented (prefixLen < 32 logic)
+   *
+   * <p>4 => match because cidrRange is IPANY_ADDRESS (0.0.0.0)
+   *
+   * <p>8 => exact match
+   */
+  private static int getIpAddressMatchScore(InetAddress localAddress, CidrRange cidrRange) {
+    if (cidrRange.getPrefixLen() == 32) {
+      try {
+        InetAddress cidrAddr = InetAddress.getByName(cidrRange.getAddressPrefix());
+        if (cidrAddr.isAnyLocalAddress()) {
+          return 4;
+        }
+        if (cidrAddr.equals(localAddress)) {
+          return 8;
+        }
+        return -1;
+      } catch (UnknownHostException e) {
+        logger.log(Level.SEVERE, "cidrRange address parsing", e);
+      }
+    }
+    // TODO(sanjaypujare): implement CIDR logic to match prefixes if needed
+    return 0;
+  }
+
+  /** Shutdown this instance and release resources. */
+  public void shutdown() {
+    logger.log(Level.INFO, "Shutdown");
+    if (xdsClient != null) {
+      xdsClient.shutdown();
+    }
+    if (timeService != null) {
+      timeServiceResource.close(timeService);
+    }
+  }
+
+  private static final class TimeServiceResource
```

Comment:
IMO, if `syncContext` comes from the caller side, this should also come from the caller side.

Reply:
But the "caller" is this class as far as XdsClientImpl is concerned. If you go up the call chain (all the way up to `XdsServerBuilder`) no other component either needs or creates a `ScheduledExecutorService` so this seems to be the right place.

Reply:
The `AbstractServerImplBuilder` defines/uses `DEFAULT_EXECUTOR_POOL` which cannot be used here because it's not a `ScheduledExecutorService` (and probably not accessible). There is no other "caller-provided" `ScheduledExecutorService` we can use unless this wrapper class creates it for `XdsClientImpl`.

Reply:
Same thing apply to syncContext. It would be very weird that they are created separately.